### PR TITLE
Use YAML for data-driven tests

### PR DESCRIPTION
**Pull request by elazarg**
Using YAML files for tests should be easier than (and complementary to) the current approach of compiling ELF files.

Proposed format:
```
test-case: re-assign immediate

pre: ["r1.value=5", "r1.type=number"]

code:
  <start>: |
    r1 = 0

post:
  - r1.type=number
  - r1.value=0
---
```

Each file is a test suite, each case describes assumed pre-invariant and expected post-invariant in (hopefully) exactly the same format as the output of the `-v` flag. 

Comparison between expected and actual is performed between the sets of string, avoiding more sophisticated equivalences, so we use the numerical domain as little as possible.

In the future, the format should be extended to deal with e.g. maps, platforms and any other kind of parameter.

The inspiration for this method is [the mypy project](https://github.com/python/mypy) (except they are using a home-brewed file format).

This PR is incomplete, but can already be used to write useful tests. We need a better, more verbose test framework. I like [pytest](https://docs.pytest.org/en/6.2.x/) (which is used by mypy) but I hesitate to add too much Python into the project.

### CODE REVIEW

**Comment by caballa** [comment]
Super useful feature!

File: src/ebpf_yaml.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    bool result = true;
+    for (const TestCase& test_case: read_suite(path)) {
+        result = result && bool(run_yaml_test_case(test_case));
+    }
+    return result;
+}
+
+void foreach_suite(const std::string& path, std::function<void(const TestCase&)> f) {
+    for (const TestCase& test_case: read_suite(path)) {
+        // print(prog, std::cout, {});
```
Comment (dthaler) [inline src/ebpf_yaml.cpp:157]: What's this comment for?  Can it be removed?

File: src/parse_constraints.hpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <vector>
+#include <set>
+#include <string>
+#include <regex>
```
Comment (dthaler) [inline src/parse_constraints.hpp:9]: alphabetize header includes?

File: scripts/.check-license.ignore [hunk 3]
```diff
@@ -5,6 +5,7 @@
 .*\.md$
 .*\.patch$
 .*\.pdf$
+.*\.yaml$
```
Comment (dthaler) [inline scripts/.check-license.ignore:8]: YAML does support comments just like bash does, so I'd recommend reverting this line and adding the license to the top of the yaml files

File: test-data/single-instruction-assignment.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+---
```
Comment (dthaler) [inline test-data/single-instruction-assignment.yaml:1]:
```suggestion
# Copyright (c) Prevail Verifier contributors.
# SPDX-License-Identifier: MIT
---
```

**Review by dthaler — APPROVED** [review]

**Pull request merged**